project setup

PawelM17 opened the pull request:
- create initial project structure,
- configure database connection,

PawelM17 committed:
- create initial project structure,
- configure database connection,

karolk92 commented on pom.xml:
```diff
@@ -0,0 +1,50 @@
+            <version>42.5.0</version>
+        </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
```
do czego potrzebujesz tej zaleznosci?

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/ResourceController.java:
```diff
@@ -0,0 +1,25 @@
+    private final ResourceService resourceService;
+
+    @PostMapping("/save")
+    public void saveResource (@RequestBody ResourceEntity resourceEntity){
```
przerob takie miejsca jak to zeby wartstwa perzystencji (encja) nie przenikala do frontu, unika sie takiego podejscia, podaj tu np. ResourceRequest ktory bedzie zawieral takie same pola jak encja a nastepnie przemapuj builderem (builder pattern) na obiekt bazodanowy (encja)

karolk92 replied:
1. moznaby tu zwrocic id zapisanego obiektu, aby front mogl wykonywac operacja przy uzyciu identyfikatora zasobu
2. nie powinienes tworzyc sciezki urla (patha) opisujacej usuwanie, tworzenie, edycja, odczyt - eng. CRUD (create, read, update, delete), od tego masz metody ktore Ci to charakteryzuja (POST, GET, PUT, DELETE)

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/ResourceController.java:
```diff
@@ -0,0 +1,25 @@
+        resourceService.saveResource(resourceEntity);
+    }
+    @DeleteMapping("/delete")
+    public void deleteResource (@RequestBody ResourceEntity resourceEntity){
```
1. usuwanie powinno odbywac sie poprzed podanie ID obiektu do usuniecia, a nie calego obiektu
2. nie powinienes tworzyc sciezki urla (patha) opisujacej usuwanie, tworzenie, edycja, odczyt - eng. CRUD (create, read, update, delete), od tego masz metody ktore Ci to charakteryzuja (POST, GET, PUT, DELETE)

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/UserController.java:
```diff
@@ -0,0 +1,28 @@
+
+    @PostMapping("/save")
+    public void saveUser(@RequestBody UserEntity userEntity){
+        userService.saveUser(userEntity);
```
zwroc id usera (zeby front mogl przy jego uzyciu robic operacje tj. usuwanie, edycja, odczyt)

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/UserController.java:
```diff
@@ -0,0 +1,28 @@
+        userService.saveUser(userEntity);
+    }
+    @DeleteMapping("/delete")
+    public void deleteUser (@RequestBody UserEntity userEntity){
```
tak jak przy resource /delete zbedne, metoda HTTP mowi ci ze to delete

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/UserController.java:
```diff
@@ -0,0 +1,28 @@
+        userService.deleteUser(userEntity);
+    }
+    @PutMapping("/update")
+    public void updateUserNickName(@RequestBody UserEntity userEntity){
```
jw. /update zbedne, jesli aktualizujesz konkretna rzecz (nick) to wystarczy ze tylko to pole podasz jako body, nie podawaj calego obiektu jesli aktualizujesz jedno pole

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/persistance/ResourceRepository.java:
```diff
@@ -0,0 +1,11 @@
+import pl.gispartner.ResourceManagerApp.model.ResourceEntity;
+
+@Repository
+public interface ResourceRepository extends JpaRepository<ResourceEntity, Long> {
```
fyi - jesli rozszerzasz JpaRepository to taka klasa automatycznie jest czytana przez kontekst springa i @Repository nie jest niezbedne, mimo to fajnie jest to zostawic dla przejrzystosci kodu

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/persistance/UserRepository.java:
```diff
@@ -0,0 +1,16 @@
+public interface UserRepository extends JpaRepository<UserEntity, Long> {
+    @Modifying
+    @Transactional
+    @Query(value = "UPDATE UserEntity SET nickName = :nickName WHERE userId = :userId")
```
@Transactional w tym miejscu zbedne - metoda albo sie wykona albo nie, doczytaj o tej adnotacji

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/ResourceService.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final ResourceRepository resourceRepository;
+
+    public void saveResource(ResourceEntity resourceEntity) {
```
przerob to tak aby nie przekazywac z frontu ResourceEntity tylko np. ResourceRequest ktory nastepnie przemapujesz na ResourceEntity (przez builder)

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/UserService.java:
```diff
@@ -0,0 +1,23 @@
+    private final UserRepository userRepository;
+
+    public void saveUser (UserEntity user){
+        userRepository.save(new UserEntity(user.getUserId(), user.getNickName(), user.getName(), user.getLastName()));
```
tutaj balagan - teoretycznie moglbys po prostu przekazac encje usera z frontu, ale przerob tak jak pisalem wyzej

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/UserService.java:
```diff
@@ -0,0 +1,23 @@
+    public void saveUser (UserEntity user){
+        userRepository.save(new UserEntity(user.getUserId(), user.getNickName(), user.getName(), user.getLastName()));
+    }
+    public void deleteUser (UserEntity user){
```
odstepy pomiedzy metodami!

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/persistance/UserRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+@Repository
+public interface UserRepository extends JpaRepository<UserEntity, Long> {
+    @Modifying
```
nie jestem pewien czy tej adnotacji (@Modyfing) uzywa sie w taki sposob, sprawdz

karolk92 reviewed (changes requested):
garsc przydatnych informacji na poczatek :)

maciejDun commented on src/main/java/pl/gispartner/ResourceManagerApp/service/UserService.java:
```diff
@@ -0,0 +1,23 @@
+        userRepository.delete(user);
+    }
+    public void updateUserNickName(Long userId, String newUserNickName){
+        userRepository.updateUserNickName(userId, newUserNickName);
```
A może tu podać cała encje z frontu

karolk92 replied:
@maciejDun encja nie powinna byc widoczna dla frontu

maciejDun commented on src/main/java/pl/gispartner/ResourceManagerApp/model/UserEntity.java:
```diff
@@ -0,0 +1,28 @@
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long userId;
+    private String nickName;
+    private String name;
```
firstName dla porządku?

maciejDun commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceEntity.java:
```diff
@@ -0,0 +1,26 @@
+//    private UserEntity userEntity; TODO
+//    private Date creationDate; TODO
+//    private Date modificationDate; TODO
+//    private String resourceType; TODO
```
Komentarzy nie wrzucamy w pr, na przyszłość może bardziej

PawelM17 committed:
- create Data Transfer Object,
- remove unnecessary URL paths,
- modification of methods,

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceDto.java:
```diff
@@ -0,0 +1,12 @@
+@AllArgsConstructor
+public class ResourceDto {
+
+    private Long resourceId;
```
zostaw po prostu id, jest w obiekcie resourceDto wiec to samo przez sie mowi ze to jest id resourca

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceEntity.java:
```diff
@@ -10,6 +10,7 @@
+@Builder
 public class ResourceEntity {
 
     @Id
```
resourceId  zmien na id - skoro jest w resource to wiadomo ze to id zasobu

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceMapper.java:
```diff
@@ -0,0 +1,14 @@
+@Component
+public class ResourceMapper {
+
+    public ResourceEntity mapToEntity(ResourceDto resourceDto) {
```
metoda powinna byc statyczna, nic nie modyfikuje, wykonuje tylko operacje. Dzieki temu bedziesz mogl sie do niej odwolac poprzez ResourceMapper.mapToEntity

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceMapper.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
nie ma potrzeby zeby mapper byl beanem

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/UserDto.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+public class UserDto {
+
+    private Long userId;
```
userId >id

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
to nie powinien byc component

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+@Component
+public class UserMapper {
+
+    public UserEntity mapToEntity(UserDto userDto) {
```
metoda statyczna

PawelM17 committed:
- improve code readability

PawelM17 committed:
- added entity created and modified date

PawelM17 committed:
- added entity type,
- improved updateUserName method;

PawelM17 committed:
- add @OneToMany and @ManyToOne relations,
- add metadata field in resources classes,
- add updateResourceName and updateJsonData methods.

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/ResourceController.java:
```diff
@@ -0,0 +1,36 @@
+        resourceService.deleteResource(resourceId);
+    }
+
+    @PutMapping("/name/{resourceId}")
```
w tej sytuacji jak zlozysz sobie urla to bedziesz mial /api/resource/name/{id} a to nie jest id imienia tylko zasobu. Zamien na /api/resource/{id}/name - w pozostalych miejscach zrob analogicznie tj. dla usera i w polach dla name i jsonData

karolk92 replied:
przekazalbym tez usera tutaj w jakis sposob i dodal walidacje czy user jest ownerem. Unikniesz sytuacji kiedy inny uzytkownik zmodyfikuje twoj zasob

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/ResourceController.java:
```diff
@@ -0,0 +1,36 @@
+        resourceService.updateResourceName(resourceId, newResourceName);
+    }
+
+    @PutMapping("/data/{resourceId}")
```
jw

karolk92 replied:
tutaj tez przekaz usera i zwaliduj

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/AuditableBase.java:
```diff
@@ -0,0 +1,28 @@
+@Setter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public class AuditableBase {
```
bardzo ladnie ze wydzielone

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceMapper.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public static String saveToString(JsonNode jsonData) {
+        ObjectMapper objectMapper = new ObjectMapper();
```
objectMapper zrob jako beana ktorego tylko wstrzykniesz do tego ResourceMappera. Zrob sobie katalog /configuration i tam zdefiniuj Beana. ResourceMappera oczywiscie przerob tak aby byl komponentem springowym

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/ResourceService.java:
```diff
@@ -0,0 +1,50 @@
+        UserEntity userEntity = userRepository.getReferenceById(resourceDto.getOwnerId());
+        resourceEntity.setUser(userEntity);
+        resourceEntity.setJsonData(ResourceMapper.saveToString(resourceDto.getJsonData()));
+        ResourceEntity created = resourceRepository.save(resourceEntity);
```
jakis enter/pusta linia dla czytelnosci

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/ResourceService.java:
```diff
@@ -0,0 +1,50 @@
+    public void updateJsonData(Long resourceId, JsonNode newJsonData) {
+        resourceRepository.updateJsonData(resourceId, newJsonData);
+        ResourceEntity resourceEntity = resourceRepository.getReferenceById(resourceId);
+        resourceEntity.setModifiedDate(new Date());
```
mozesz tez uzyc LocalDate.now()

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/UserService.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public void deleteUser(Long userId) {
+        userRepository.deleteById(userId);
```
te walidacje o ktorym pisalem (czy odpowiedni uzytkownik najlepiej zrobic w serwisie), tutaj np. moglbys zwalidowac ze jesli usuwasz siebie to nie ma problemu, a zeby usunac innego uzytkownika to musisz byc SUPER_USER

karolk92 reviewed (changes requested):
ogolnie rzecz biorac widac duze postepy, jestem zadowolony :)

PawelM17 committed:
- improve URL paths,
- add validating methods.

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/api/ResourceController.java:
```diff
@@ -19,18 +19,33 @@ public Long saveResource(@RequestBody ResourceDto resourceDto) {
-    public void deleteResource(@PathVariable("resourceId") Long resourceId) {
-        resourceService.deleteResource(resourceId);
+    public String deleteResource(@PathVariable("resourceId") Long resourceId, @RequestHeader Long userId) {
+        if (resourceService.validateResourceOwner(userId, resourceId)) {
```
przetlumacz sobie to zdanie - "jesli zwaliduj wlasciciela zasobu", brzmi kiepsko, nie? jesli uzywasz w ifie to nazwe metody trzeba zmienic na np. `isResourceOwnerValid` wiedy w ifie bedzie `if (isResourceOwnerValid(userId, resourceId)` czyli "jesli jest poprawny wlasciciel zasobu to zrob to".
Tego pierwszego wariantu moglbys uzyc gdybys nic nie zwracal z ifa, a wewnatrz metody rzucal wyjatek.

karolk92 replied:
jak dla mnie - najlepiej byloby tak: (jesli juz nie chcesz tego do serwisu przenosic - bo to bylaby jeszcze lepsza opcja)
`        if (!resourceService.isResourceOwnerValid(userId, resourceId)) {
               return "This operation cannot be performed";  
        }
          resourceService.deleteResource(resourceId);  
          return "Changes have been successfully saved";`

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/model/ResourceEntity.java:
```diff
@@ -35,5 +37,16 @@ public class ResourceEntity extends AuditableBase {
     private String jsonData;
 
-
+    @Override
```
lombok ma adnotacje od eq and hash

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/ResourceService.java:
```diff
@@ -47,4 +52,10 @@ public void updateJsonData(Long resourceId, JsonNode newJsonData) {
     }
 
+    public boolean validateResourceOwner(Long userId, Long resourceId) {
+        ResourceEntity resourceEntity = resourceRepository.getReferenceById(resourceId);
```
na przyszlosc z tym`getReferenceById` raczej uzywaj `findById`, nigdy w kodzie nie widzialem zeby ktos tego uzyl wiec lepiej uzywac tego co wiekszosc. Nie mowie ze to podejscie jest trafne w kazdym przypadku ale wtedy na pewno twoj kod bedzie czytelniejszy dla innych

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/ResourceService.java:
```diff
@@ -47,4 +52,10 @@ public void updateJsonData(Long resourceId, JsonNode newJsonData) {
+    public boolean validateResourceOwner(Long userId, Long resourceId) {
+        ResourceEntity resourceEntity = resourceRepository.getReferenceById(resourceId);
+        UserEntity userEntity = userRepository.getReferenceById(userId);
+        return Objects.equals(userId, resourceEntity.getUser().getId()) || userEntity.getUserType().equals(SUPER_USER);
```
raz uzywasz Objects.equals a raz equals, badz konsekwentny i uzyj jednej opcji

karolk92 commented on src/main/java/pl/gispartner/ResourceManagerApp/service/UserService.java:
```diff
@@ -31,4 +34,9 @@ public void updateUserName(Long userId, String newUserName) {
+
+    public boolean validateUser(Long id, Long userId) {
+        UserEntity userEntity = userRepository.getReferenceById(id);
+        return Objects.equals(id, userId) || userEntity.getUserType().equals(SUPER_USER);
```
jw

PawelM17 committed:
- add GetMapping,
- improve validating methods,
- improve updating methods,
- add AppConfiguration class,
- improve ResourceMapper class.

karolk92 reviewed (approved):
elegancko, etap 1 za nami! :)
